### PR TITLE
[KYUUBI #1536] Add session conf advisor

### PR DESCRIPTION
**Pull request by ulysses-you**
### _Why are the changes needed?_

closes https://github.com/apache/incubator-kyuubi/issues/1536

### _How was this patch tested?_
- [x] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [ ] [Run test](https://kyuubi.apache.org/docs/latest/develop_tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala [hunk 1]
```diff
@@ -1172,6 +1172,13 @@ object KyuubiConf {
       .checkValues(OperationLanguages.values.map(_.toString))
       .createWithDefault(OperationLanguages.SQL.toString)
 
+  val SESSION_CONF_ADVISOR: ConfigEntry[String] =
+    buildConf("session.conf.advisor")
+      .doc("Choose a custom config advisor for session, the default implementation return empty.")
```
Comment (SteNicholas) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:1177]: Is it better that the default implementation return null?

**Review by SteNicholas — COMMENTED** [review]
@ulysses-you, thanks for your contribution. I have left minor comments for this pull request.

File: kyuubi-server/pom.xml [hunk 2]
```diff
@@ -31,6 +31,12 @@
     <packaging>jar</packaging>
 
     <dependencies>
+        <dependency>
+            <groupId>org.apache.kyuubi</groupId>
+            <artifactId>kyuubi-plugin</artifactId>
```
Comment (yaooqinn) [inline kyuubi-server/pom.xml:36]: kyuubi-server-plugin

File: kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSessionImpl.scala [hunk 3]
```diff
@@ -45,11 +50,33 @@ class KyuubiSessionImpl(
...
+      case ("kyuubi.engine.pool.size.threshold", _) =>
+      case (key, value) =>
+        sessionConf.set(key, value)
+        userSpecifiedConf.put(key, value)
+    }
+
+    val sessionConfMap = ImmutableMap.copyOf[String, String](sessionConf.getAll.asJava)
+    sessionManager.sessionConfAdvisor.getConfSuggestion(
+      user,
+      sessionConfMap).asScala.foreach {
```
Comment (yaooqinn) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSessionImpl.scala:67]: NPE?

File: kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSessionManager.scala [hunk 4]
```diff
@@ -27,13 +27,15 @@ import org.apache.kyuubi.credentials.HadoopCredentialsManager
...
 import org.apache.kyuubi.operation.KyuubiOperationManager
+import org.apache.kyuubi.plugin.{PluginLoader, SessionConfAdvisor}
 
 class KyuubiSessionManager private (name: String) extends SessionManager(name) {
 
   def this() = this(classOf[KyuubiSessionManager].getSimpleName)
 
   val operationManager = new KyuubiOperationManager()
   val credentialsManager = new HadoopCredentialsManager()
+  lazy val sessionConfAdvisor: SessionConfAdvisor = PluginLoader.loadSessionConfAdvisor(conf)
```
Comment (yaooqinn) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSessionManager.scala:38]: don't be lazy?

File: kyuubi-plugin/src/main/java/org/apache/kyuubi/plugin/SessionConfAdvisor.java [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+import java.util.HashMap;
+import java.util.Map;
+
+/** Provide the session configuration according to the user and session configuration. */
+public interface SessionConfAdvisor {
+  /** The returned conf is used if it is not existed in session conf */
+  default Map<String, String> getConfSuggestion(String user, Map<String, String> sessionConf) {
+    return new HashMap<>();
+  }
```
Comment (yaooqinn) [inline kyuubi-plugin/src/main/java/org/apache/kyuubi/plugin/SessionConfAdvisor.java:28]: on second thought, we'd better remove this API to keep it unambiguous

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala (hunk 1 above)
Comment (ulysses-you) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:1177]: changed to optional

File: kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSessionImpl.scala [hunk 6]
```diff
@@ -45,11 +50,30 @@ class KyuubiSessionImpl(
...
+      sessionConfMap)
+    if (confOverlay != null) {
+      confOverlay.asScala.foreach {
+        case (key, value) =>
+          sessionConf.set(key, value)
+          userSpecifiedConf.put(key, value)
+      }
+    }
+
+    userSpecifiedConf.asScala.toMap
```
Comment (yaooqinn) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSessionImpl.scala:76]: simply add ?

```scala
private[kyuubi] val optimizedConf: Map[String, String] = {
    val confOverlay = sessionManager.sessionConfAdvisor.getConfSuggestion(
      user,
      normalizedConf.asJava)
    normalizedConf ++ confOverlay.asScala
  }
```

File: .github/workflows/style.yml [hunk 7]
```diff
@@ -41,6 +41,12 @@ jobs:
         with:
           distribution: zulu
           java-version: 8
+      - name: install kyuubi with the latest commit
```
Comment (yaooqinn) [inline .github/workflows/style.yml:44]: RELATED?

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala [hunk 8]
```diff
@@ -1172,6 +1172,13 @@ object KyuubiConf {
       .checkValues(OperationLanguages.values.map(_.toString))
       .createWithDefault(OperationLanguages.SQL.toString)
 
+  val SESSION_CONF_ADVISOR: OptionalConfigEntry[String] =
+    buildConf("session.conf.advisor")
+      .doc("Choose a custom config advisor for session, the default implementation return empty.")
```
Comment (yaooqinn) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:1177]: let's detail the doc of the config, e.g. what's happening while using this plugin, not just duplicates the config name with some words reordered, besides `the default implementation return empty` is redundant too as we will have a `<td>` for default value.

Comment (yaooqinn) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:1177]: Additionally, we need a separate doc for the plugin. there is nowhere to find the relation between SessionConfAdvisor and this config

File: .github/workflows/style.yml (hunk 7 above)
Comment (ulysses-you) [reply .github/workflows/style.yml:44]: the new added failed will fail in style check workflow if we do not install first. So I changed the ga in this PR

File: kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSessionManager.scala (hunk 4 above)
Comment (yaooqinn) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSessionManager.scala]: i see,

File: .github/workflows/style.yml (hunk 7 above)
Comment (yaooqinn) [inline .github/workflows/style.yml:44]:
```suggestion
      - name: Install
```

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala [hunk 9]
```diff
@@ -1172,6 +1172,16 @@ object KyuubiConf {
       .checkValues(OperationLanguages.values.map(_.toString))
       .createWithDefault(OperationLanguages.SQL.toString)
 
+  val SESSION_CONF_ADVISOR: OptionalConfigEntry[String] =
+    buildConf("session.conf.advisor")
+      .doc("A config advisor plugin for Kyuubi Server. This plugin can provide some custom " +
+        "configs for different user or session configs and overwrite the session config before " +
+        "open a new session. This config value should be a class which is a child of " +
+        "org.apache.kyuubi.plugin.SessionConfAdvisor.")
```
Comment (yaooqinn) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:1180]: zero-arg constructor?

File: kyuubi-server-plugin/src/main/java/org/apache/kyuubi/plugin/SessionConfAdvisor.java [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+package org.apache.kyuubi.plugin;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/** Provide the session configuration according to the user and session configuration. */
+public interface SessionConfAdvisor {
+  /** The returned conf will overwrite the session conf */
+  default Map<String, String> getConfOverlay(String user, Map<String, String> sessionConf) {
+    return new HashMap<>();
```
Comment (yaooqinn) [inline kyuubi-server-plugin/src/main/java/org/apache/kyuubi/plugin/SessionConfAdvisor.java:27]: Nit:Collections.emptyMap();?

File: kyuubi-server/pom.xml [hunk 11]
```diff
@@ -31,6 +31,12 @@
     <packaging>jar</packaging>
 
     <dependencies>
+        <dependency>
```
Comment (yaooqinn) [inline kyuubi-server/pom.xml:34]: to L40?

File: kyuubi-server/src/main/scala/org/apache/kyuubi/plugin/PluginLoader.scala [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+ */
+
+package org.apache.kyuubi.plugin
+
+import scala.util.control.NonFatal
+
+import org.apache.kyuubi.KyuubiException
+import org.apache.kyuubi.config.KyuubiConf
+
+object PluginLoader {
```
Comment (yaooqinn) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/plugin/PluginLoader.scala:25]: nit:private[kyuubi]

File: kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSessionManager.scala (hunk 4 above)
Comment (yaooqinn) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSessionManager.scala:39]: can we init immediately when `initialize`

File: .github/workflows/style.yml [hunk 13]
```diff
@@ -41,6 +41,12 @@ jobs:
         with:
           distribution: zulu
           java-version: 8
+      - name: install
```
Comment (yaooqinn) [inline .github/workflows/style.yml:44]: nit: Install？

File: kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSessionImpl.scala [hunk 14]
```diff
@@ -45,8 +47,20 @@ class KyuubiSessionImpl(
...
+      normalizedConf ++ confOverlay.asScala
+    } else {
+      warn(s"the server plugin return null value for user: $user, ignore it")
+      normalizedConf
+    }
+  }
+
   // TODO: needs improve the hardcode
-  normalizedConf.foreach {
+  optimizedConf.foreach {
```
Comment (yaooqinn) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSessionImpl.scala:63]: we shall also update it at L107

Comment (ulysses-you) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSessionImpl.scala:63]: ah, good catch !

File: kyuubi-server/pom.xml [hunk 15]
```diff
@@ -37,6 +37,12 @@
             <version>${project.version}</version>
         </dependency>
 
+        <dependency>
```
Comment (yaooqinn) [inline kyuubi-server/pom.xml:40]: cp this to codecov

**Review by yaooqinn — APPROVED** [review]

**Comment by ulysses-you** [comment]
thanks, merging to master

**Pull request closed without merging**